### ISSUE TITLE
Buffer pooling for marten

### ISSUE THREAD

**Issue by Scooletz** [JasperFx/marten#640]
Below you can find a description of providing buffer pooling for Marten.

---

Currently, `EventStreamAppender` uses `JsonBodies` func to append bodies as an array of strings ([here](https://github.com/JasperFx/marten/blob/2f18d09fa2034cbc647f48a74bbf3bbb8ea51116/src/Marten/Events/EventStreamAppender.cs#L37)).
According to `Npgsql` driver, it's possible to pass a `char[]` instead of string and use the parameter size to limit the length of the appended data ([here](https://github.com/npgsql/npgsql/blob/4ef74fa78cffbb4b1fdac00601d0ee5bff5e242b/src/Npgsql/TypeHandlers/TextHandler.cs#L269-L276). 

We can't use it as we pass the array parameter, but we could 
- enable passing `ArraySegment<char>` as `NpgsqlDbType.Jsonb` in Npgsql driver first
- enable serializers to serialize to a predefined `char[]` (+ possibly reallocation if needed)
- pass result `ArraySegment<char>` below.

Having this, we could leverage that for other operations, like storing documents (`DocJsonBodyArgument` uses serializer as well). Actually, **we could start with docs first** as they issue a single `Jsonb` parameter (hence, we can use `Size`) and later, decide do we want it for events.

I'm aware that these allocations are short living. On the other hand we can't tell is it worth it before trying it out.

What do you think?

**Comment by jeremydmiller** [comment]
@Scooletz It's a little bit related to https://github.com/JasperFx/marten/issues/610. I know that we're doing more allocations than is strictly necessary internally in Marten. I chickened out of trying to do this for 1.2, but I'd like to see this happen. 

Hey, in a busy system, this might be well worth the effort. I got stopped last time on the identity map abstractions, but that's a problem that we can beat.
